assets/js/modules/analytics-4/utils/data-mock.js

```diff
@@ -0,0 +1,558 @@
+ *
+ * @since n.e.x.t
+ *
+ * @param {Array.<Object>}        rows    Array of rows to sort.
+ * @param {Array.<Object>}        metrics Array of report metrics.
+ * @param {Object|Array.<Object>} orderby Sorting options.
+ * @return {Array.<Object>} Sorted rows.
+ */
+function sortRows( rows, metrics, orderby ) {
+	let sorted = rows;
+
+	const orders = castArray( orderby );
+	for ( const order of orders ) {
+		const direction = order?.sortOrder === 'DESCENDING' ? -1 : 1;
+		const index = metrics.findIndex(
+			( metric ) => getMetricKey( metric ) === order?.fieldName
+		);
+		if ( index < 0 ) {
+			continue;
+		}
+
+		sorted = sorted.sort( ( a, b ) => {
+			let valA = parseFloat( a.metrics[ index ]?.values?.[ 0 ] );
+			if ( Number.isNaN( valA ) ) {
+				valA = 0;
+			}
+
+			let valB = parseFloat( b.metrics[ index ]?.values?.[ 0 ] );
+			if ( Number.isNaN( valB ) ) {
+				valB = 0;
+			}
+
+			return ( valA - valB ) * direction;
+		} );
+	}
+
+	return sorted;
+}
+
+/**
+ * Generates date range.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} startDate The start date.
+ * @param {string} endDate   The end date.
+ * @return {Array.<string>} An array with dates.
+ */
+function generateDateRange( startDate, endDate ) {
+	const dates = [];
+
+	const currentDate = stringToDate( startDate );
+	const end = stringToDate( endDate );
+
+	while ( currentDate.getTime() <= end.getTime() ) {
+		dates.push(
+			currentDate.toISOString().split( 'T' )[ 0 ].replace( /\D/g, '' )
+		);
+		currentDate.setDate( currentDate.getDate() + 1 );
+	}
+
+	return dates;
+}
+
+/**
+ * Generates mock data for Analytics 4 reports.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object} args Report options.
+ * @return {Array.<Object>} An array with generated report.
+ */
+export function getAnalytics4MockResponse( args ) {
+	invariant(
+		isPlainObject( args ),
+		'report options are required to generate a mock response.'
+	);
+	invariant(
+		isValidDateString( args.startDate ),
+		'a valid startDate is required.'
+	);
+	invariant(
+		isValidDateString( args.endDate ),
+		'a valid endDate is required.'
+	);
+
+	const originalSeedValue = faker.seedValue;
+	const argsHash = parseInt(
+		md5( args.url || 'http://example.com' ).substring( 0, 8 ),
+		16
+	);
+
+	// We set seed for every data mock to make sure that the same arguments get the same report data.
+	// It means that everyone will have the same report data and will see the same widgets in the storybook.
+	// This approach gives us additional flexibility to control randomness on a per widget basis.
+	if ( ! Number.isNaN( argsHash ) ) {
+		faker.seed( argsHash );
+	}
+
+	const data = {
+		rowCount: 0,
+		rows: [],
+		metadata: {
+			currencyCode: 'USD',
+			timeZone: 'America/Los_Angeles',
+		},
+		kind: 'analyticsData#runReport',
+	};
+
+	const { compareStartDate, compareEndDate } = args;
+	const hasDateRange = compareStartDate && compareEndDate;
+
+	const validMetrics = ( args.metrics || [] ).filter(
```

Comment:
I think we should pass the full type names in `args.metrics`, as discussed elsewhere. So `args.metrics` should contain `TYPE_INTEGER`, `TYPE_STRING` etc. I think this would be preferable as it keeps things closer to the real report data, even if it means the interface is slightly different to the UA mock utility.

Reply:
@techanvil Can you be more explicit here please? 😅  I would prefer to have the metrics to be in the same format as the API, i.e no need to have the types in there.

Reply:
Hi @asvinb, my apologies, my comment was complete gibberish, I had misread the code. Please ignore!